File: c++/525. Contiguous Array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int res = 0;
        int n = nums.size();
        int sum = 0;
        unordered_map<int, int> um;
        um[0] = -1;
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i]==1)
                sum += 1;
            else
                sum += -1;
            if (um.count(sum) == 0)
                um[sum] = i;
            else {
                int last = um[sum];
                res = max(res, i - last);
            }
        }

        return res;
    }
};

File: go/src/leetcode/525. Contiguous Array/525. Contiguous Array.go
package main

/**
思路：转化为和=target最长非空子数组长度
 */
func findMaxLength(nums []int) int {
	var ans int
	var s int
	mapper := make(map[int]int)
	mapper[0] = -1

	for i, num := range nums {
		if num == 0 {
			s--
		} else {
			s++
		}

		if last, ok := mapper[s]; !ok {
			mapper[s] = i
		} else {
			ans = max(ans, i-last+1)
		}
	}
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
